java.lang.RuntimeException: Unable to start activity ComponentInfo

I'm facing some issues with the latest version of the package (`flutter_appauth: 4.2.0`):

It seems like some activity for `RedirectUriReceiverActivity` is missing / cannot be found - anyone with the same issue or any idea for the reason?


**LOGS:**

```
E/AndroidRuntime(12050): 	at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:3270)
E/AndroidRuntime(12050): 	at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:3409)
E/AndroidRuntime(12050): 	at android.app.servertransaction.LaunchActivityItem.execute(LaunchActivityItem.java:83)
E/AndroidRuntime(12050): 	at android.app.servertransaction.TransactionExecutor.executeCallbacks(TransactionExecutor.java:135)
E/AndroidRuntime(12050): 	at android.app.servertransaction.TransactionExecutor.execute(TransactionExecutor.java:95)
E/AndroidRuntime(12050): 	at android.app.ActivityThread$H.handleMessage(ActivityThread.java:2016)
E/AndroidRuntime(12050): 	at android.os.Handler.dispatchMessage(Handler.java:107)
E/AndroidRuntime(12050): 	at android.os.Looper.loop(Looper.java:214)
E/AndroidRuntime(12050): 	at android.app.ActivityThread.main(ActivityThread.java:7356)
E/AndroidRuntime(12050): 	at java.lang.reflect.Method.invoke(Native Method)
E/AndroidRuntime(12050): 	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:492)
E/AndroidRuntime(12050): 	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:930)
E/AndroidRuntime(12050): Caused by: java.lang.IllegalStateException: You need to use a Theme.AppCompat theme (or descendant) with this activity.
E/AndroidRuntime(12050): 	at androidx.appcompat.app.AppCompatDelegateImpl.createSubDecor(AppCompatDelegateImpl.java:852)
E/AndroidRuntime(12050): 	at androidx.appcompat.app.AppCompatDelegateImpl.ensureSubDecor(AppCompatDelegateImpl.java:815)
E/AndroidRuntime(12050): 	at androidx.appcompat.app.AppCompatDelegateImpl.initWindowDecorActionBar(AppCompatDelegateImpl.java:556)
E/AndroidRuntime(12050): 	at androidx.appcompat.app.AppCompatDelegateImpl.getSupportActionBar(AppCompatDelegateImpl.java:543)
E/AndroidRuntime(12050): 	at androidx.appcompat.app.AppCompatDelegateImpl.invalidateOptionsMenu(AppCompatDelegateImpl.java:1226)
E/AndroidRuntime(12050): 	at androidx.appcompat.app.AppCompatActivity.invalidateOptionsMenu(AppCompatActivity.java:314)
E/AndroidRuntime(12050): 	at androidx.activity.ComponentActivity.invalidateMenu(ComponentActivity.java:553)
E/AndroidRuntime(12050): 	at androidx.activity.-$$Lambda$6vUCJMIem6_TsIeNg4KPj4HWSJA.run(Unknown Source:2)
E/AndroidRuntime(12050): 	at androidx.core.view.MenuHostHelper.addMenuProvider(MenuHostHelper.java:116)
E/AndroidRuntime(12050): 	at androidx.activity.ComponentActivity.addMenuProvider(ComponentActivity.java:531)
E/AndroidRuntime(12050): 	at androidx.fragment.app.FragmentActivity$HostCallbacks.addMenuProvider(FragmentActivity.java:736)
E/AndroidRuntime(12050): 	at androidx.fragment.app.FragmentManager.attachController(FragmentManager.java:2784)
E/AndroidRuntime(12050): 	at androidx.fragment.app.FragmentController.attachHost(FragmentController.java:117)
E/AndroidRuntime(12050): 	at androidx.fragment.app.FragmentActivity.lambda$init$3$FragmentActivity(FragmentActivity.java:140)
E/AndroidRuntime(12050): 	at androidx.fragment.app.-$$Lambda$FragmentActivity$euPNEtWNfVdMY89Jt5kWt_WEHqw.onContextAvailable(Unknown Source:2)
E/AndroidRuntime(12050): 	at androidx.activity.contextaware.ContextAwareHelper.dispatchOnContextAvailable(ContextAwareHelper.java:99)
E/AndroidRuntime(12050): 	at androidx.activity.ComponentActivity.onCreate(ComponentActivity.java:352)
E/AndroidRuntime(12050): 	at androidx.fragment.app.FragmentActivity.onCreate(FragmentActivity.java:217)
E/AndroidRuntime(12050): 	at net.openid.appauth.RedirectUriReceiverActivity.onCreate(RedirectUriReceiverActivity.java:49)
E/AndroidRuntime(12050): 	at android.app.Activity.performCreate(Activity.java:7802)
E/AndroidRuntime(12050): 	at android.app.Activity.performCreate(Activity.java:7791)
E/AndroidRuntime(12050): 	at android.app.Instrumentation.callActivityOnCreate(Instrumentation.java:1299)
E/AndroidRuntime(12050): 	at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:3245)
E/AndroidRuntime(12050): 	... 11 more
I/Process (12050): Sending signal. PID: 12050 SIG: 9
Lost connection to device.
```

Happens to me as well. Tested it with various emulators and physical devices running different android SDKs, but the result remains unfortunately the same.

```logdart
E/AndroidRuntime(32012): Caused by: java.lang.IllegalStateException: You need to use a Theme.AppCompat theme (or descendant) with this activity.
E/AndroidRuntime(32012): 	at androidx.appcompat.app.AppCompatDelegateImpl.createSubDecor(AppCompatDelegateImpl.java:852)
E/AndroidRuntime(32012): 	at androidx.appcompat.app.AppCompatDelegateImpl.ensureSubDecor(AppCompatDelegateImpl.java:815)
E/AndroidRuntime(32012): 	at androidx.appcompat.app.AppCompatDelegateImpl.initWindowDecorActionBar(AppCompatDelegateImpl.java:556)
E/AndroidRuntime(32012): 	at androidx.appcompat.app.AppCompatDelegateImpl.getSupportActionBar(AppCompatDelegateImpl.java:543)
E/AndroidRuntime(32012): 	at androidx.appcompat.app.AppCompatDelegateImpl.invalidateOptionsMenu(AppCompatDelegateImpl.java:1226)
E/AndroidRuntime(32012): 	at androidx.appcompat.app.AppCompatActivity.invalidateOptionsMenu(AppCompatActivity.java:314)
E/AndroidRuntime(32012): 	at androidx.activity.ComponentActivity.invalidateMenu(ComponentActivity.java:553)
E/AndroidRuntime(32012): 	at androidx.activity.-$$Lambda$6vUCJMIem6_TsIeNg4KPj4HWSJA.run(Unknown Source:2)
E/AndroidRuntime(32012): 	at androidx.core.view.MenuHostHelper.addMenuProvider(MenuHostHelper.java:116)
E/AndroidRuntime(32012): 	at androidx.activity.ComponentActivity.addMenuProvider(ComponentActivity.java:531)
E/AndroidRuntime(32012): 	at androidx.fragment.app.FragmentActivity$HostCallbacks.addMenuProvider(FragmentActivity.java:736)
E/AndroidRuntime(32012): 	at androidx.fragment.app.FragmentManager.attachController(FragmentManager.java:2784)
E/AndroidRuntime(32012): 	at androidx.fragment.app.FragmentController.attachHost(FragmentController.java:117)
E/AndroidRuntime(32012): 	at androidx.fragment.app.FragmentActivity.lambda$init$3$FragmentActivity(FragmentActivity.java:140)
E/AndroidRuntime(32012): 	at androidx.fragment.app.-$$Lambda$FragmentActivity$euPNEtWNfVdMY89Jt5kWt_WEHqw.onContextAvailable(Unknown Source:2)
E/AndroidRuntime(32012): 	at androidx.activity.contextaware.ContextAwareHelper.dispatchOnContextAvailable(ContextAwareHelper.java:99)
E/AndroidRuntime(32012): 	at androidx.activity.ComponentActivity.onCreate(ComponentActivity.java:352)
E/AndroidRuntime(32012): 	at androidx.fragment.app.FragmentActivity.onCreate(FragmentActivity.java:217)
E/AndroidRuntime(32012): 	at net.openid.appauth.RedirectUriReceiverActivity.onCreate(RedirectUriReceiverActivity.java:49)
E/AndroidRuntime(32012): 	at android.app.Activity.performCreate(Activity.java:7994)
E/AndroidRuntime(32012): 	at android.app.Activity.performCreate(Activity.java:7978)
E/AndroidRuntime(32012): 	at android.app.Instrumentation.callActivityOnCreate(Instrumentation.java:1309)
E/AndroidRuntime(32012): 	at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:3422)
E/AndroidRuntime(32012): 	... 11 more
I/Process (32012): Sending signal. PID: 32012 SIG: 9
Lost connection to device.


```

Also started to run into this issue recently. I am pretty surprised at the proposed solutions tbh. We are not using `local_auth`

The relevant parts of the logs are:
```
E/AndroidRuntime(12050): Caused by: java.lang.IllegalStateException: You need to use a Theme.AppCompat theme (or descendant) with this activity.
[...]
E/AndroidRuntime(12050): 	at net.openid.appauth.RedirectUriReceiverActivity.onCreate(RedirectUriReceiverActivity.java:49)
E/AndroidRuntime(12050): 	at android.app.Activity.performCreate(Activity.java:7802)
E/AndroidRuntime(12050): 	at android.app.Activity.performCreate(Activity.java:7791)
E/AndroidRuntime(12050): 	at android.app.Instrumentation.callActivityOnCreate(Instrumentation.java:1299)
E/AndroidRuntime(12050): 	at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:3245)
E/AndroidRuntime(12050): 	... 11 more
I/Process (12050): Sending signal. PID: 12050 SIG: 9
```

This means that for some reason the `RedirectUriReceiverActivity` from the android library [here](https://github.com/openid/AppAuth-Android/blob/master/library/java/net/openid/appauth/RedirectUriReceiverActivity.java) is started without an AppCompat theme. Which is surprising as the `RedirectUriReceiverActivity` extends correctly from `AppCompatActivity` and in our Flutter app `MainActivity` is also extending the correct `FlutterFragmentActivity` as well as using the correct styles with `Theme.AppCompat`.

I am still scratching my head to what changed for us, as the plugin did work recently. The only potential issue I see is that the plugin uses the base `Activity` instead of any `AppCompat` [here](https://github.com/MaikuB/flutter_appauth/blob/f999c01d6e1811baf85b9a6b55ca2dfa89ce71f1/flutter_appauth/android/src/main/java/io/crossingthestreams/flutterappauth/FlutterAppauthPlugin.java#L72). And also does not use `AppCompat.startActivityForResult` [here](https://github.com/MaikuB/flutter_appauth/blob/f999c01d6e1811baf85b9a6b55ca2dfa89ce71f1/flutter_appauth/android/src/main/java/io/crossingthestreams/flutterappauth/FlutterAppauthPlugin.java#L352C22-L352C44). (Which is now additionally deprecated)

So my last guess so far would be that there might be sort of erasure as only the superclass `startActivityForResult` is called instead of the AppCompat version and therefore the Theme is not overwritten with the correct Theme.AppCompat. I will try to make some time to validate this.

Anyway regarding the proposed fixes, imo this is the way: In your manifest overwrite the redirect activity with an appropriate appcompat theme. 

```
        <activity
            android:name="net.openid.appauth.RedirectUriReceiverActivity"
            android:exported="true"
            android:theme="@style/Theme.AppCompat.Light.NoActionBar"
            tools:node="replace">
            <intent-filter>
                <action android:name="android.intent.action.VIEW"/>
                <category android:name="android.intent.category.DEFAULT"/>
                <category android:name="android.intent.category.BROWSABLE"/>
                <data android:scheme="your-scheme-check-docu-for-info"/>
            </intent-filter>
        </activity>
```

And on a personal note, I do think this is an issue in the plugin even if I can not pinpoint it yet. Maybe it would help to also update the example app to some up2date version of flutter.

--- a/flutter_appauth/CHANGELOG.md
+++ b/flutter_appauth/CHANGELOG.md
@@ -1,6 +1,7 @@
 ## [7.0.0-dev.3]
 
 * **Breaking change** all methods have now been made to return non-nullable types
+* [Android] updated plugin to specify a theme for the `RedirectUriReceiverActivity` from the AppAuth Android SDK
 * [iOS][macOS] bumped AppAuth iOS dependency to 1.7.5
 * Updated API docs with more details
 * Updated readme with more details on essential knowledge and links to OAuth 2.0 specifications

--- a/flutter_appauth/CHANGELOG.md
+++ b/flutter_appauth/CHANGELOG.md
@@ -1,7 +1,7 @@
 ## [7.0.0-dev.3]
 
 * **Breaking change** all methods have now been made to return non-nullable types
-* [Android] updated plugin to specify a theme for the `RedirectUriReceiverActivity` from the AppAuth Android SDK
+* [Android] updated plugin to specify `Theme.AppCompat.Translucent.NoTitleBar` as the theme for the `RedirectUriReceiverActivity` from the AppAuth Android SDK. This is to fix a crash raised with issues [#362](https://github.com/MaikuB/flutter_appauth/issues/362) and [#515](https://github.com/MaikuB/flutter_appauth/issues/515)
 * [iOS][macOS] bumped AppAuth iOS dependency to 1.7.5
 * Updated API docs with more details
 * Updated readme with more details on essential knowledge and links to OAuth 2.0 specifications

--- a/flutter_appauth/android/src/main/AndroidManifest.xml
+++ b/flutter_appauth/android/src/main/AndroidManifest.xml
@@ -1,3 +1,9 @@
 <manifest xmlns:android="http://schemas.android.com/apk/res/android"
-  package="io.crossingthestreams.flutterappauth">
+  package="io.crossingthestreams.flutterappauth"
+  xmlns:tools="http://schemas.android.com/tools">
+  <application>
+    <activity android:name="net.openid.appauth.RedirectUriReceiverActivity"
+      android:theme="@style/Theme.AppCompat.Translucent.NoTitleBar"
+      tools:replace="android:theme" />
+  </application>
 </manifest>

